json/json.go

```diff
@@ -0,0 +1,172 @@
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/sourcegraph/zoekt"
+	"github.com/sourcegraph/zoekt/query"
+)
+
+// defaultTimeout is the maximum amount of time a search request should
+// take. This is the same default used by Sourcegraph.
+const defaultTimeout = 20 * time.Second
+
+func JSONServer(searcher zoekt.Searcher) http.Handler {
+	s := jsonSearcher{searcher}
+	mux := http.NewServeMux()
+	mux.HandleFunc("/search", s.jsonSearch)
+	mux.HandleFunc("/list", s.jsonList)
+	return mux
+}
+
+type jsonSearcher struct {
+	Searcher zoekt.Searcher
+}
+
+type jsonSearchArgs struct {
+	Q    string
+	Opts *zoekt.SearchOptions
+}
+
+type jsonSearchReply struct {
+	Result *zoekt.SearchResult
+}
+
+type jsonListArgs struct {
+	Q    string
+	Opts *zoekt.ListOptions
+}
+
+type jsonListReply struct {
+	List *zoekt.RepoList
+}
+
+func (s *jsonSearcher) jsonSearch(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	w.Header().Add("Content-Type", "application/json")
+
+	if req.Method != "POST" {
+		jsonError(w, http.StatusMethodNotAllowed, "Only POST is supported")
+		return
+	}
+
+	searchArgs := jsonSearchArgs{}
+	err := json.NewDecoder(req.Body).Decode(&searchArgs)
+	if err != nil {
+		jsonError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+	if searchArgs.Q == "" {
+		jsonError(w, http.StatusBadRequest, "missing query")
+		return
+	}
+	if searchArgs.Opts == nil {
+		searchArgs.Opts = &zoekt.SearchOptions{}
+	}
+
+	query, err := query.Parse(searchArgs.Q)
+	if err != nil {
+		jsonError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	// Set a timeout if the user hasn't specified one.
+	if searchArgs.Opts.MaxWallTime == 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, defaultTimeout)
+		defer cancel()
+	}
+
+	if err := CalculateDefaultSearchLimits(ctx, query, s.Searcher, searchArgs.Opts); err != nil {
+		jsonError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	searchResult, err := s.Searcher.Search(ctx, query, searchArgs.Opts)
+	if err != nil {
+		jsonError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	json.NewEncoder(w).Encode(jsonSearchReply{searchResult})
+}
+
+func jsonError(w http.ResponseWriter, statusCode int, err string) {
+	w.WriteHeader(statusCode)
+	json.NewEncoder(w).Encode(struct{ Error string }{Error: err})
+}
+
+// Calculates and sets heuristic defaults on opts for various upper bounds on
+// the number of matches when searching, if none are already specified. The
+// defaults are derived from opts.MaxDocDisplayCount, so if none is set, there
+// is no calculation to do.
+func CalculateDefaultSearchLimits(ctx context.Context,
+	q query.Q,
+	searcher zoekt.Searcher,
+	opts *zoekt.SearchOptions) error {
+	maxResultDocs := opts.MaxDocDisplayCount
+	if maxResultDocs != 0 &&
+		opts.ShardMaxMatchCount == 0 &&
+		opts.ShardMaxImportantMatch == 0 {
+		// This is a special mode of Search that _only_ calculates ShardFilesConsidered and bails ASAP.
```

Comment:
nit: can we flip this around to reduce nesting?

Would this work?
```
	if opts.MaxDocDisplayCount == 0 || opts.ShardMaxMatchCount != 0 || opts.ShardMaxImportantMatch != 0 {
		return nil
	}

	maxResultDocs := opts.MaxDocDisplayCount
```

Reply:
Done.